### PR TITLE
get_token consistently requires at least one scope

### PR DESCRIPTION
**Pull request by chlowell**
Calling a credential's `get_token` method with no argument is equivalent to requesting an access token from Azure AD for no particular scope. Azure AD rejects such requests as malformed. Allowing `get_token()` also doesn't make sense in the Azure SDK because the caller presumably wants a token to authorize some request, which implies the caller knows details of that request including the required scope of authorization.

`get_token()` should therefore raise an error, and it typically does. However, in some cases it can succeed due to unintended consequences of a credential's internal use of `msal`. To prevent confusing failure and misleading success (see #10243 for an example), this change defines the behavior of `get_token()`: it raises in all cases.

### CODE REVIEW

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

File: sdk/identity/azure-identity/CHANGELOG.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Release History
 
 ## 1.4.0b2 (Unreleased)
-
+- Calling any credential's `get_token` method with no scope raises `ValueError`
```
Comment (johanste) [inline sdk/identity/azure-identity/CHANGELOG.md]: I would rephrase this to start with which method is changed and why it was changed. E.g. 

`TokenCredentials.get_token` requires at least one scope to be provided. While calling the method with no scopes may have worked in earlier versions, the underlying behavior was actually undefined and very likely to change as soon as more tokens were added to the underlying shared cache.

File: sdk/identity/azure-identity/azure/identity/_internal/decorators.py [hunk 2]
```diff
@@ -9,6 +9,19 @@
 from azure.core.exceptions import ClientAuthenticationError
 
 
+def at_least_one_scope(get_token):
```
Comment (johanste) [inline sdk/identity/azure-identity/azure/identity/_internal/decorators.py]: By having a decorator, you only change two lines of code into one line of code at the cost of (some) runtime cost (calling additional methods). I would do the parameter check in the method body instead.

File: sdk/identity/azure-identity/CHANGELOG.md (hunk 1 above)
Comment (lmazuel) [reply sdk/identity/azure-identity/CHANGELOG.md]: I agree with Johan, I would expect the few impacted people to search for this info in the ChangeLog, it's what I do when pytest/requests/etc. is breaking me, looking for the ChangeLog to understand why (before creating an issue). Just seeing "yeah now we raise" is not the best experience we can provide.

File: sdk/identity/azure-identity/CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,11 @@
 # Release History
 
 ## 1.4.0b2 (Unreleased)
+- All `get_token` methods consistently require at least one scope argument,
+raising an error when none is passed. Although `get_token()` may sometimes
+have succeeded in prior versions, it couldn't do so consistently because its
+behavior was undefined, and dependened on the credential's type and internal
+state. ([#10366](https://github.com/Azure/azure-sdk-for-python/pull/10366))
```
Comment (johanste) [inline sdk/identity/azure-identity/CHANGELOG.md]: Can we link to the/an issue instead? The issue can be marked as resolved by this PR.

Comment (chlowell) [reply sdk/identity/azure-identity/CHANGELOG.md]: I linked the PR because its description focuses on the rationale behind this change and links to the motivating issue as an example of what goes wrong without it. I think that issue's more narrow focus on a specific Graph + Key Vault scenario makes it a worse place to start for someone wanting more information about this change.

But if you aren't persuaded sure, we can link to the issue 😸

Comment (johanste) [reply sdk/identity/azure-identity/CHANGELOG.md]: I generally prefer to link to issues since they are inherently talking about a single issue. Pointing to the PR gets a bit recursive/meta since we are discussing how to document the change in the PR comments :). 

If you feel like the issue is too specific, feel free to open a new issue that outlines the specific context.

File: sdk/identity/azure-identity/CHANGELOG.md [hunk 4]
```diff
@@ -1,6 +1,11 @@
 # Release History
 
 ## 1.4.0b2 (Unreleased)
+- All `get_token` methods consistently require at least one scope argument,
+raising an error when none is passed. Although `get_token()` may sometimes
+have succeeded in prior versions, it couldn't do so consistently because its
+behavior was undefined, and dependened on the credential's type and internal
+state. ([#10243](https://github.com/Azure/azure-sdk-for-python/pull/10243))
```
Comment (lmazuel) [inline sdk/identity/azure-identity/CHANGELOG.md]: picky: GH is nice with us, but technically your URL contains "pull" even if it's not :)
https://github.com/Azure/azure-sdk-for-python/issues/10243
